core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java

```diff
@@ -1071,12 +1462,70 @@ private void mergeAndLogEntry(FileEntry entry, int mergeCounter) {
          throw new PersistenceException("Could not add new merged entry", e);
       }
    }
+
    @Override
-   public void purge(Executor threadPool, final PurgeListener task) {
-      long now = timeService.wallClockTime();
+   public Publisher<MarshallableEntry<K, V>> purgeExpired() {
+      UnicastProcessor<MarshallableEntry<K, V>> processor = UnicastProcessor.create();
+      blockingManager.runBlocking(() -> blockingPurgeExpired(processor), "sfs-purgeExpired");
+      return processor;
+   }
+
+   private void blockingPurgeExpired(UnicastProcessor<MarshallableEntry<K, V>> processor) {
+      try {
+         long now = timeService.wallClockTime();
+         for (int segment = 0; segment < actualNumSegments; segment++) {
+            List<KeyValuePair<Object, FileEntry>> entriesToPurge;
+            long stamp = resizeLock.readLock();
+            try {
+               Map<K, FileEntry> segmentEntries = getSegmentEntries(segment);
+               if (segmentEntries == null)
+                  continue;
+
+               entriesToPurge = collectExpiredEntries(now, segmentEntries);
+            } finally {
+               resizeLock.unlockRead(stamp);
```

Comment:
Don't we need to hold this read lock until we are done with the FileEntry objects as  before? TBH I am not clear what the locks a protecting anymore sadly :( Or should we lock the FileEntry when we are about to purge them?

Reply:
No, the resizeLock is only used to protect the `entries` array so segments can't be removed while we are iterating them. Each `FileEntry` then has it's own locking, so they are not freed while someone is reading from them.

Reply:
> Each FileEntry then has it's own locking, so they are not freed while someone is reading from them.

Yeah, but my second sentence is shouldn't we be doing that? I see it is unlocking the FileEntry in `purgeExpiredEntries` when invoking `readFromDisk` but it is never locked. I was expecting the FileEntry to be locked in `collectExpiredEntries` while holding the resizeLock.

Reply:
It looks like you are somewhat, but not completely, getting around it by removing the entry from the ConcurrentMap and the fact that `free` now waits for readers to complete. Unfortunately, there is still a window where a reader can see the FileEntry due to `readFromDisk` decrementing readers without incrementing it in `blockingPurgeExpired`. This means we could decrement a legitimate reader and they may see a FileEntry in a possibly incorrect state as it can be overwritten. I would think this would be a very small window and rarely occur, but we should probably have a JIRA just to not lose it.

Reply:
:+1:

I thought about making the lock release in `readFromDisk` conditional, but I decided it wouldn't be worth the complexity.